### PR TITLE
Empty antler-run with eos-vm and eosiolib linked

### PR DESCRIPTION
**Pull request by dimas1185**
#79 

Challenges:
* `eosiolib` haven't been built with `gcc`. it was only built with `cdt-cpp` before. However when we included `eosiolib` into `antler-run` we got chicken and egg problem: all tools are built with default compiler (gcc in case of Ubuntu) and then `eosiolib` built with those tools. Solution was to slightly change `eosiolib` to be built with gcc. So now we have two versions of `eosiolib` lib for different purposes. 
* Another challenge was to built 'eosiolib' with standard library (previously it was only built with custom standard lib that is located in `libraries/libc++`. The reason to use standard library is because of we add `eos-vm` as well and there were a lot to fix to get it along with custom `libc++`. `Libc++` lacks `fstream` and other stuff that is needed by `eos-vm`.
* native library was also added to `antler-run`. Issue with it was that we can't have our own `_start` with standard library. So this code was split to avoid this function included for `antler-run` binary. `native` library now also has two native builds with different toolchains (gcc and cdt-cpp).

Note: there is a decision to remove Mac support. This PR maintains Mac support in order to makes changes distinct. I'm planning to make another PR to remove Mac support. 

## API Changes
- [ ] API Changes

## Documentation Additions
- [ ] Documentation Additions

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -10,3 +10,6 @@
 [submodule "libraries/native/softfloat"]
 	path = libraries/native/softfloat
 	url = https://github.com/AntelopeIO/berkeley-softfloat-3
+[submodule "tools/external/eos-vm"]
+	path = tools/external/eos-vm
+	url = git@github.com:AntelopeIO/eos-vm
```
Comment (spoonincode) [inline .gitmodules:15]: public repos should really be https:// instead

**Review by spoonincode — CHANGES_REQUESTED** [review]

File: libraries/eosiolib/core/eosio/datastream.hpp [hunk 2]
```diff
@@ -16,6 +16,8 @@
 #include <optional>
 #include <variant>
 
+#include <string.h>
```
Comment (larryk85) [inline libraries/eosiolib/core/eosio/datastream.hpp:19]: use `include <cstring>`.

File: libraries/eosiolib/capi/eosio/types.h [hunk 3]
```diff
@@ -32,15 +32,25 @@ typedef uint64_t capi_name;
 /**
  * EOSIO Public Key. K1 and R1 keys are 34 bytes.  Newer keys can be variable-sized
  */
-struct __attribute__((deprecated("newer public key types cannot be represented as a fixed size structure", "char[]")))
+struct 
+#ifdef __GNUC__
+__attribute__((deprecated("newer public key types cannot be represented as a fixed size structure, use char[] instead")))
```
Comment (larryk85) [inline libraries/eosiolib/capi/eosio/types.h:37]: Why have the `ifdef`?

File: libraries/native/elf_crt.s [hunk 4]
```diff
@@ -16,60 +8,4 @@ _start:
...
-	mov 8(%rdi), %rbp
-	mov 16(%rdi), %r12
-	mov 24(%rdi), %r13
-	mov 32(%rdi), %r14
-	mov 40(%rdi), %r15
-	mov 48(%rdi), %rdx
-	mov %rdx, %rsp
-	mov 56(%rdi), %rdx
-	jmp *%rdx
+   syscall
```
Comment (larryk85) [inline libraries/native/elf_crt.s:11]: We shouldn't edit the `elf_crt.s` as we won't need the `crt` for compiling as a shared object.

File: libraries/native/mac_utils.s [hunk 5]
```diff
@@ -0,0 +1,60 @@
+.global ____putc
```
Comment (larryk85) [inline libraries/native/mac_utils.s:1]: We don't need any mac os support as this time.

File: libraries/native/macho_crt.s [hunk 6]
```diff
@@ -11,60 +7,4 @@ start:
...
-	mov 8(%rdi), %rbp
-	mov 16(%rdi), %r12
-	mov 24(%rdi), %r13
-	mov 32(%rdi), %r14
-	mov 40(%rdi), %r15
-	mov 48(%rdi), %rdx
-	mov %rdx, %rsp
-	mov 56(%rdi), %rdx
-	jmp *%rdx
+   syscall
```
Comment (larryk85) [inline libraries/native/macho_crt.s:10]: We can remove this macho_crt.s for the time being as we don't support mac os directly for the foreseeable future.

File: tools/antler-run/file-utils.hpp [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    uninitialized,
+    non_elf_other,
+    elf_object,
+    elf_executable,
+    elf_shared_object,
+    elf_core_dump,
+    wasm
+};
+
+file_type get_file_type(const char* path) {
```
Comment (larryk85) [inline tools/antler-run/file-utils.hpp:23]: make this `inline`.

File: tools/antler-run/file-utils.hpp [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        if (wasm_magic == eosio::vm::magic)
+            return file_type::wasm;
+        
+        return file_type::non_elf_other;
+    }
+
+    return file_type::uninitialized;
+}
+
+std::string file_type_str(file_type t) {
```
Comment (larryk85) [inline tools/antler-run/file-utils.hpp:58]: make this `inline`.

**Review by larryk85 — CHANGES_REQUESTED** [review]
Small changes, other than that looking good.

File: libraries/eosiolib/capi/eosio/types.h (hunk 3 above)
Comment (dimas1185) [reply libraries/eosiolib/capi/eosio/types.h:37]: clang's `deprecated` has two parameters (one for type name, another for suggested replacement).
gcc has one parameter of single string

File: libraries/native/elf_crt.s (hunk 4 above)
Comment (dimas1185) [reply libraries/native/elf_crt.s:11]: correct, I added comments in makefile about that. however this code is used in few intrinsics (like `eosio_assert`) that we will have to write runner implementation. So I planned to stitch it back later, when we will agree on intrinsics architecture.

File: libraries/native/macho_crt.s (hunk 6 above)
Comment (dimas1185) [reply libraries/native/macho_crt.s:10]: I wanted to do this in a separate commit, as there are few more places with mac os specifics.

File: tools/antler-run/file-utils.hpp (hunk 7 above)
Comment (dimas1185) [reply tools/antler-run/file-utils.hpp:23]: will do.
p.s. not a big deal anyway but internet suggests:
`The compiler declines the request of inlining a function if it contains any switch or go-to statements`

File: libraries/eosiolib/capi/eosio/types.h (hunk 3 above)
Comment (ScottBailey) [reply libraries/eosiolib/capi/eosio/types.h:37]: clang's feature is nicer, though it's optional, it seems worth keeping.

File: tools/antler-run/file-utils.hpp (hunk 7 above)
Comment (ScottBailey) [reply tools/antler-run/file-utils.hpp]: Resolved.

**Review by ScottBailey — APPROVED** [review]
lgtm

**Review by larryk85 — APPROVED** [review]
Does this break anything currently in use?  If it does then let's make a branch called `next_release` and merge to that.

**Comment by dimas1185** [comment]
@larryk85 tests are fine and binary workflow supposed to be unchanged, I made two different implementations - `crt.cpp` and `crt_lib.cpp`. later is only used for shared library

**Pull request merged**